### PR TITLE
Refs [TASK][34438]Handle event between each fragment

### PR DESCRIPTION
**Pull request by Hoang-Kelvin**
## Related Tickets
- [#Task 34438: Handle event between each fragment](https://edu-redmine.sun-asterisk.vn/issues/34438)
## WHAT
- Handle click event in 5 main fragments
- Fix bug and controll backstack
## Evidence (Screenshot or Video)
image

## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/project/gamedb/base/BaseActivity.kt [hunk 1]
```diff
@@ -25,7 +25,8 @@ abstract class BaseActivity : AppCompatActivity(), BaseView {
         fragmentContainer.visibility = View.VISIBLE
         viewPagerFragment.visibility = View.GONE
         tabLayout.visibility = View.GONE
-        supportFragmentManager.beginTransaction().replace(R.id.fragmentContainer, fragment).commit()
+        supportFragmentManager.beginTransaction().replace(R.id.fragmentContainer, fragment)
+            .addToBackStack(getString(R.string.string_fragment)).commit()
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/base/BaseActivity.kt]:
```suggestion
        supportFragmentManager.beginTransaction()
            .replace(R.id.fragmentContainer, fragment)
            .addToBackStack(getString(R.string.string_fragment))
            .commit()
```

File: app/src/main/java/com/project/gamedb/ui/more/MoreFragment.kt [hunk 2]
```diff
@@ -1,10 +1,80 @@
...
+    }
+
+    private fun initData() {
+        when (ordering) {
+            ApiConstants.GENRES -> morePresenter?.apply {
+                getInfo(id)
+                getGamesOrdered(ordering, id)
+            }
+            else -> morePresenter?.getGamesOrdered(ordering, id)
+        }
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/more/MoreFragment.kt]: 2 case thì dùng if else

File: app/src/main/java/com/project/gamedb/ui/more/MorePresenter.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import com.project.gamedb.data.model.ResultGames
+import com.project.gamedb.data.source.remote.GameRemoteRepository
+import com.project.gamedb.data.source.remote.OnDataLoadedCallback
+import java.lang.Exception
+
+class MorePresenter(
+    private val view: MoreContract.View,
+    private val repository: GameRemoteRepository
+) : MoreContract.Presenter {
+    override fun getGamesOrdered(ordering: String, query: String) {
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/more/MorePresenter.kt]:
```suggestion
) : MoreContract.Presenter {

    override fun getGamesOrdered(ordering: String, query: String) {
```

File: app/src/main/java/com/project/gamedb/ui/more/MorePresenter.kt [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        repository.getMoreOption(ordering, query, object : OnDataLoadedCallback<ResultGames>{
+            override fun onSuccess(data: ResultGames) {
+                view.showGames(data)
+            }
+
+            override fun onFailure(exception: Exception?) {
+                view.showError(exception?.message.toString())
+            }
+
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/more/MorePresenter.kt]:
```suggestion
            }
```

File: app/src/main/java/com/project/gamedb/ui/more/MorePresenter.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    private val repository: GameRemoteRepository
+) : MoreContract.Presenter {
+    override fun getGamesOrdered(ordering: String, query: String) {
+        repository.getMoreOption(ordering, query, object : OnDataLoadedCallback<ResultGames>{
+            override fun onSuccess(data: ResultGames) {
+                view.showGames(data)
+            }
+
+            override fun onFailure(exception: Exception?) {
+                view.showError(exception?.message.toString())
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/more/MorePresenter.kt:21]: nếu exception null thì show gì ?

File: app/src/main/java/com/project/gamedb/ui/more/MoreViewHolder.kt [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            setPlatform(this, getPlatform(item.gamePlatforms.toString()))
+            textGenresFeature.text = getListGenres(item.gameGenres)
+            textItemFeature?.text = item.gameName
+            gameItemFeature?.loadImage(item.gameBackgroundImage)
+            this.setOnClickListener {
+                onClickDetailsListener.openGameDetail(
+                    item.gameId.toInt(),
+                    getListGenres(item.gameGenres)
+                )
+            }
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/more/MoreViewHolder.kt]: set On Click trong init{} của ViewHolder tránh gán lại nhiều lần

File: app/src/main/java/com/project/gamedb/ui/more/MoreViewHolder.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    private val itemView: View,
+    private val onClickDetailsListener: OnClickDetailsListener
+) : BaseViewHolder<Games>(itemView) {
+
+    override fun onBind(item: Games) {
+        itemView.run {
+            setPlatform(this, getPlatform(item.gamePlatforms.toString()))
+            textGenresFeature.text = getListGenres(item.gameGenres)
+            textItemFeature?.text = item.gameName
+            gameItemFeature?.loadImage(item.gameBackgroundImage)
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/more/MoreViewHolder.kt:39]: Tại sao phải check null 2 item này

File: app/src/main/java/com/project/gamedb/ui/more/MoreViewHolder.kt [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        return platforms
+    }
+
+    private fun getListGenres(genres: List<Genres>): String {
+        val list = mutableListOf<String>()
+        for (gen in genres) {
+            list.add(gen.genresName)
+        }
+        return list.toString().replace("[", "").replace("]", "")
+    }
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/more/MoreViewHolder.kt]: dùng map, joinToString()

File: app/src/main/java/com/project/gamedb/ui/more/MoreViewHolder.kt [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        for (i in platforms.indices) {
+            image[i].setPlatform(platforms[i])
+        }
+    }
+
+    private fun getPlatform(strings: String): List<String> {
+        val platforms = mutableListOf<String>()
+        for (string in PlatformConstants.platforms) {
+            if (strings.contains(string)) platforms.add(string)
+        }
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/more/MoreViewHolder.kt]: dùng filter

File: app/src/main/java/com/project/gamedb/ui/popular/PopularFragment.kt [hunk 10]
```diff
@@ -1,41 +1,64 @@
...
+    private var mainAdapter: PopularAdapter? = null
+    private var itemAdapter: FeatureAdapter? = null
     private var mainPresenter: PopularPresenter? = null
+    private var mainCallback: OnFragmentIntegrationListener.Open? = null
 
     override val layoutResource get() = R.layout.fragment_popular
 
+    init {
+        mainAdapter = PopularAdapter(this)
+        itemAdapter = FeatureAdapter(this)
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/popular/PopularFragment.kt]: sao không gán luôn bên trên mà phải gán ở init{}

File: app/src/main/java/com/project/gamedb/ui/popular/PopularFragment.kt [hunk 11]
```diff
@@ -1,41 +1,64 @@
...
-class PopularFragment : BaseFragment(), PopularContract.View {
-    private val mainAdapter = PopularAdapter()
-    private val itemAdapter = FeatureAdapter()
+class PopularFragment : BaseFragment(), PopularContract.View, OnClickDetailsListener {
+    private var mainAdapter: PopularAdapter? = null
+    private var itemAdapter: FeatureAdapter? = null
     private var mainPresenter: PopularPresenter? = null
+    private var mainCallback: OnFragmentIntegrationListener.Open? = null
 
     override val layoutResource get() = R.layout.fragment_popular
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/popular/PopularFragment.kt:21]: override thì cho lên đầu

File: app/src/main/java/com/project/gamedb/ui/popular/PopularFragment.kt [hunk 12]
```diff
@@ -1,41 +1,64 @@
...
+        textItemLoading?.hide()
+        itemAdapter?.replaceData(games.results)
     }
 
     override fun showError(error: String) {
         context?.showToast(error, Toast.LENGTH_SHORT)
     }
 
+    override fun openGameDetail(id: Int, genres: String) {
+        mainCallback?.openNewFragment(DetailsFragment(id, genres))
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/popular/PopularFragment.kt:54]: sai convention 2.10

File: app/src/main/java/com/project/gamedb/ui/popular/PopularFragment.kt [hunk 13]
```diff
@@ -47,6 +70,14 @@ class PopularFragment : BaseFragment(), PopularContract.View {
     private fun initAdapter() {
         recyclerGameFeature.adapter = mainAdapter
         recyclerGameItem.adapter = itemAdapter
+        buttonSeeAll.setOnClickListener {
+            mainCallback?.openNewFragment(
+                MoreFragment(
+                    getString(R.string.string_order_rating),
+                    getString(R.string.string_ordering),
+                )
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/popular/PopularFragment.kt]: sai convention 2.10

File: app/src/main/java/com/project/gamedb/ultis/ViewExtentions.kt [hunk 14]
```diff
@@ -34,3 +34,11 @@ fun View.loadDrawable(context: Context, urlImage: String) {
             })
     }
 }
+
+fun View.show() {
+    this.visibility = View.VISIBLE
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ultis/ViewExtentions.kt]: remove this.

File: app/src/main/res/layout/item_recyclerview_main.xml [hunk 15]
```diff
@@ -28,6 +28,8 @@
         app:layout_constraintTop_toTopOf="parent" />
 
     <TextView
+        android:ellipsize="end"
+        android:maxLines="1"
         android:id="@+id/textGenresFeature"
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"
```
Comment (huytq-2422) [inline app/src/main/res/layout/item_recyclerview_main.xml]: format code

File: app/src/main/java/com/project/gamedb/ui/popular/PopularViewHolder.kt [hunk 16]
```diff
@@ -2,15 +2,37 @@ package com.project.gamedb.ui.popular
...
+import com.project.gamedb.data.model.Genres
 import com.project.gamedb.ultis.loadImage
 import kotlinx.android.synthetic.main.item_recyclerview_main.view.*
 
-class PopularViewHolder(private val itemView: View) : BaseViewHolder<Games>(itemView) {
+class PopularViewHolder(
+    private val itemView: View,
+    private val onClickDetailsListener: OnClickDetailsListener
+) :
+    BaseViewHolder<Games>(itemView) {
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/popular/PopularViewHolder.kt]:
```suggestion
) :   BaseViewHolder<Games>(itemView) {
```

File: app/src/main/java/com/project/gamedb/ui/popular/PopularViewHolder.kt [hunk 17]
```diff
@@ -2,15 +2,37 @@ package com.project.gamedb.ui.popular
...
         itemView.run {
+            textGenresFeature.text = getListGenres(item.gameGenres)
             textGameFeature.text = item.gameName
             imageFeature.loadImage(item.gameBackgroundImage)
+            buttonMoreInfo.setOnClickListener {
+                onClickDetailsListener.openGameDetail(
+                    item.gameId.toInt(),
+                    getListGenres(item.gameGenres)
+                )
+            }
```
Comment (huytq-2422) [inline app/src/main/java/com/project/gamedb/ui/popular/PopularViewHolder.kt:39]: setOnClick in Init{}

File: app/src/main/java/com/project/gamedb/ui/more/MorePresenter.kt (hunk 5 above)
Comment (Hoang-Kelvin) [reply app/src/main/java/com/project/gamedb/ui/more/MorePresenter.kt:21]: Nó sẽ ghi là value null at ....

File: app/src/main/java/com/project/gamedb/ui/more/MoreViewHolder.kt (hunk 7 above)
Comment (Hoang-Kelvin) [reply app/src/main/java/com/project/gamedb/ui/more/MoreViewHolder.kt:39]: Nếu dữ liệu được gửi về nhưng ng dùng đã destroy View trước đó thì check để tránh bị crash app

**Pull request merged**